Remove tax unit groupings.

MImmesberger opened the pull request:
Closes #690

lars-reimann commented:
Regarding [this comment](https://github.com/iza-institute-of-labor-economics/gettsim/pull/691#issuecomment-1971726293):

> @lars-reimann Can you have a look where the `CondaVerificationError` in the `readthedocs` check comes from?

I could not reproduce this locally. It seems to be caused by the upgrade of `sphinx-automodapi` to version 0.17.0. In previous, successful runs version 0.16.0 was used. Maybe you can add an upper bound for `sphinx-automodapi` in `environment.yml` in this PR and try again?

```txt
- sphinx-automodapi<0.17.0
```

lars-reimann commented:
That seems to have fixed the install issue at least.

MImmesberger commented:
Yes, I guess the other errors are due to this incomplete PR. Thanks!

hmgaudecker reviewed (commented):
Just a couple of remarks while browsing the changes.

lars-reimann commented on src/_gettsim/groupings.py:
```diff
@@ -123,27 +184,24 @@ def sn_id_numpy(
 
     for index, current_p_id in enumerate(p_id):
-        current_p_id_einstandspartner = p_id_ehepartner[index]
+        current_p_id_ehepartner = p_id_ehepartner[index]
```
That's to match the input variable `p_id_ehepartner`. If `einstandspartner` is correct, this change should be reverted and the input parameter changes instead.

hmgaudecker replied:
Sorry, my bad! I thought this was another function.

lars-reimann commented on src/_gettsim/aggregation_numpy.py:
```diff
@@ -17,7 +17,8 @@ def grouped_count(group_id):
     fail_if_dtype_not_int(group_id, agg_func="grouped_sum")
     fail_if_dtype_not_numeric_or_boolean(column, agg_func="grouped_sum")
-
+    if column.dtype == bool:
```
Jax already had this conversion.

MImmesberger commented on docs/geps/gep-01.md:
```diff
@@ -42,7 +42,7 @@ a nutshell and without explanations, these conventions are:
 
 1. If names need to be concatenated for making clear what a column name refers to (e.g.,
-   `arbeitsl_geld_2_vermög_freib_bg` vs. `grunds_im_alter_vermög_freib_hh`), the group
+   `arbeitsl_geld_2_vermög_freib_bg` vs. `grunds_im_alter_vermög_freib_fg`), the group
```
This was intentional but now as you mentioned it, I think the function itself should be grouped on the `eg` level. Rather, the inputs should be on the `fg` level:

```python
def grunds_im_alter_vermög_freib_eg(
    anz_erwachsene_fg: int,
    anz_kinder_fg: int,
    grunds_im_alter_params: dict,
) -> float:
...
    out = (
        grunds_im_alter_params["vermögensfreibetrag"]["adult"] * anz_erwachsene_fg
        + grunds_im_alter_params["vermögensfreibetrag"]["child"] * anz_kinder_fg
    )
    return out
```

MImmesberger commented on src/_gettsim/transfers/unterhaltsvors.py:
```diff
@@ -68,56 +103,73 @@ def unterhaltsvors_m(  # noqa: PLR0913
         out > 0
         and (alter >= unterhaltsvors_params["altersgrenze_mindesteinkommen"])
-        and (unterhaltsvorschuss_eink_m_tu < unterhaltsvors_params["mindesteinkommen"])
+        and (not _unterhaltsvorschuss_eink_above_income_threshold_fg)
```
This is just an approximation of the legal requirements because the income check is not really on the `fg` level:

Conceptually, we need to check whether the parent that receives Unterhaltsvorschuss satisfy an income threshold. This parental-level boolean should then be used on the child-level to check whether the child is eligible for Unterhaltsvorschuss. This is then again aggregated to the parent level to calculate `unterhaltsvors_m`.

I think we are not able to handle this case yet (without introducing a new input variable besides `p_id_unterhaltsvors_empf`).

hmgaudecker replied:
I'm afraid I don't follow the second paragraph just yet.
- How does that back and forth work?
- Is the check at the parent level really on the individual income? That is, if my `einstandspartner` is a millionaire, doesn't her income count?

MImmesberger replied:
I think the problem is that we have pointers from children to parents, but not from parents to children. So, my point is that we cannot properly depict this minimum income check in this PR. Using `fg` as a grouping works as long as no child earns more than the minimum income threshold.

> And can't we determine p_id_unterhaltsvors_empf endogenously based on kindesunterhalt and parental ids?

Yes that's true, forgot about that.

hmgaudecker replied:
Lets see whether I understood correctly. We have

| p_id | fg_id | alter | p_id_unterhaltsvors_empf | einkommen_m | unterhalt |
| ---- | ----- | ----- | ------------------------ | ----------- | --------- |
| 0    | 0     | 35    | -1                       | 100         | 0         |
| 1    | 0     | 15    | 0                        | 0           | 0         |
| 2    | 1     | 35    | -1                       | 1000        | 0         |
| 3    | 1     | 15    | 2                        | 0           | 0         |
| 4    | 0     | 13    | 2                        | 0           | 500       |

and want to get at:

| p_id | unterhaltsvors_kind |
| ---- | ------------------- |
| 0    | 0                   |
| 1    | 0                   |
| 2    | 0                   |
| 3    | 250                 |
| 4    | 0                   |

because first parent does not qualify for income check and last kid receives too much Unterhalt.

The problem is, of course, that if we work on row-based data and vectorize, we can't get at: *(using pandas syntax)*

```py
data.loc["p_id_unterhaltsvors_empf", "einkommen"]
```

However, we could create a variables like this upfront with the whole aggregation stuff. In this case, we are not aggregating by any values, but we are only using pointers (this is the unique end of the m:1 relationship between parents and children). 

*In case this is correct:* This does not seem hard to do in case we don't have the machinery just yet. @lars-reimann, could you have a look at that? Might be worth doing a separate PR for that, using the above as the input and the intermediate variable:

| p_id | _unterhaltsvors_empf_einkommen_m |
| ---- | ------------------- |
| 0    | nan                   |
| 1    | 100.0                   |
| 2    | nan                   |
| 3    | 1000.0                 |
| 4    | 1000.0                   |

as the output of a test?

MImmesberger replied:
Yes, that's exactly the problem.

hmgaudecker replied:
@lars-reimann, this probably got buried in the flurry of Marvin's activity -- could you have a look at my suggestion two messages above? I am thinking of a function `get_value_from_other_p_id(variable, others_p_id)` or so.

hmgaudecker replied:
Will be fixed in #708

hmgaudecker commented on src/_gettsim/config.py:
```diff
@@ -99,6 +99,15 @@ def set_array_backend(backend: str):
         "description": "Spouses filing taxes jointly or individuals.",
         "potentially_endogenous": True,
     },
+    "ehe": {
```
Just making a note here to see how often we'll need it in the end (apologies that my writing was confused over in #690). 

At least these days, it probably won't ever be relevant to work with aggregates at this level (sn/eg will be the relevant groupings). Might be different 20 years ago though.

The problem with it is that the aggregate is confusing for single persons when thinking about it intuitively. So there would be some value in doing without. Again, nothing to be done right now, just leaving this as a note.

hmgaudecker commented on src/_gettsim/config.py:
```diff
@@ -99,6 +99,15 @@ def set_array_backend(backend: str):
+    },
+    "eg": {
+        "name": "Einstandsgemeinschaft / Einstandspartner",
+        "description": "Couples that live together but are not necessarily married.",
```
```suggestion
        "description": "A couple whose members are deemed to be responsible for each other.",
```

MImmesberger commented:
Note: Currently, we need to specifiy `p_id_ehepartner` and `p_id_einstandspartner` separately. Might make sense to set `p_id_einstandspartner` automatically if `p_id_ehepartner` is specified.

hmgaudecker commented:
I don't see a point in that. May save a user a line of code and create confusion if we silently modify one of her variables.

Still, very good point! Just a different conclusion: please add a data consistency check on this.

MImmesberger commented on src/_gettsim_tests/test_data/arbeitsl_geld_2/2005/hh_id_7.yaml:
```diff
@@ -72,11 +68,11 @@ inputs:
       - 0.0
       - 0.0
-    soli_st_y_tu:
+    soli_st_y_sn:
```
I think the test here doesn't make much sense because it assumes that three adults>25 have a Bedarfsgemeinschaft together. Should we get rid of it then?

hmgaudecker replied:
Yes, that does not make any sense. Can't see much of the file on the phone, but if there are no comments in it that should make us reconsider, just delete it.

MImmesberger replied:
No comments, and it's hard to tell what the test is actually about. Will delete it, thanks.

MImmesberger commented:
Some things that I have encountered:
1. I have renamed `eigener_bedarf_gedeckt` to `eigenbedarf_gedeckt` to adhere with the naming rules.
2. `p_id_einstandspartner` is 1 character too long. Maybe rename all of these grouping pointers to `pid_...`? Or go with `p_id_ep`?
3. Betreuungskosten are now aggregated via `p_id`. Currently I use the user-specified Betreuungskosten-Träger ID `p_id_betreuungsk_trä` (at the character limit) as the `p_id_to_aggregate_by`. Alternatively, we could use `p_id_kindergeld_empf`.
4. The Wohngeld Vorrang check in `wohngeld_vorrang_hh` is wrong (as previously discussed, I added a ToDo for that that links to #690 ).

Additionally two questions regarding Grundsicherung im Alter:
1. Currently, Grundsicherung im Alter is only payed out if all individuals in the household are retired. There is a ToDo without a linked issue. I guess we don't touch this part in this PR?
2. `grunds_im_alter_m_eg` uses `arbeitsl_geld_2_regelbedarf_m_bg` as relevant income. I don't know whether the correct grouping should be `arbeitsl_geld_2_regelbedarf_m_eg`, but then again, there is the question on how to split up the ALG2 benefits.

hmgaudecker commented:
> I have renamed `eigener_bedarf_gedeckt` to `eigenbedarf_gedeckt` to adhere with the naming rules.

Is that the correct term from the law / commonplace? Sounds funny, but that does not mean anything.

hmgaudecker commented:
> * `p_id_einstandspartner` is 1 character too long. Maybe rename all of these grouping pointers to `pid_...`? Or go with `p_id_ep`?

Let's just disable the test. Change is around the corner and we should not spend time on this anymore.

hmgaudecker commented:
> * Betreuungskosten are now aggregated via `p_id`. Currently I use the user-specified Betreuungskosten-Träger ID `p_id_betreuungsk_trä` (at the character limit) as the `p_id_to_aggregate_by`. Alternatively, we could use `p_id_kindergeld_empf`.

See previous thing, just go to `p_id_betreuungsk_träger`. But we probably won't loose much if using the Kindergeld version? I'd probably favor the latter but that'll be your call.

hmgaudecker commented:
> 1. Currently, Grundsicherung im Alter is only payed out if all individuals in the household are retired. There is a ToDo without a linked issue. I guess we don't touch this part in this PR?

No, that can remain a to-do, this PR is humongous already. Maybe @ChristianZimpelmann remembers the background?

> 1. `grunds_im_alter_m_eg` uses `arbeitsl_geld_2_regelbedarf_m_bg` as relevant income. I don't know whether the correct grouping should be `arbeitsl_geld_2_regelbedarf_m_eg`, but then again, there is the question on how to split up the ALG2 benefits.

We might have to repeat those calculations, unfortunately. But if you can find a workaround for now, no need to solve it at this very moment!

MImmesberger commented:
> Is that the correct term from the law / commonplace? Sounds funny, but that does not mean anything.

Yes, the Arbeitsagentur [uses it](https://www.arbeitsagentur.de/datei/dok_ba015899.pdf)

MImmesberger commented:
> We might have to repeat those calculations, unfortunately. But if you can find a workaround for now, no need to solve it at this very moment!

The current implementation may be a bad approximation if there are retirees that have children in their Bedarfsgemeinschaft. Even then, the current implementation might be correct, but I couldn't find anything online. But I think this is not too common, so we can ignore it for now.

MImmesberger commented on src/_gettsim/interface.py:
```diff
@@ -241,18 +241,6 @@ def _process_and_check_data(data):
-    if ("tu_id" in data) and ("hh_id" in data):
-        assert (
-            not data["tu_id"].groupby(data["hh_id"]).std().max() > 0
-        ), "We currently allow for only one tax unit within each household"
```
I think this ToDo is done with this PR, right? Else we can revert this commit.

ChristianZimpelmann commented:
I think this PR could also close #89

> 1. Currently, Grundsicherung im Alter is only payed out if all individuals in the household are retired. There is a ToDo without a linked issue. I guess we don't touch this part in this PR?

We were not sure how to handle cases in which one person in the Bedarfsgemeinschaft is retired, but another person is unemployed and searching for a job. We opted for this simplified version, but we should reconsider at some point. I now opened an issue for it: #703 Can you link it in the code?

MImmesberger commented:
The tax unit grouping is now removed in the code and any mention of it in the docs is gone.

I slightly adjusted the tutorials by manually grouping the GETTSIM outputs on the `hh` level where `tu` groupings were used before. I think this keeps the tutorial as simple as possible.

hmgaudecker commented on src/_gettsim/groupings.py:
```diff
@@ -33,6 +39,61 @@ def bg_id_numpy(
     return numpy.asarray(result)
 
 
+def eg_id_numpy(
```
Just flagging missing test coverage -- probably a fluke, but maybe this was overseen.

hmgaudecker commented on src/_gettsim/groupings.py:
```diff
@@ -33,6 +39,61 @@ def bg_id_numpy(
     return numpy.asarray(result)
 
 
+def eg_id_numpy(
```
Do we have checks for whether eg / ehe have size 1 or 2?

MImmesberger replied:
Added issue #707

hmgaudecker commented on src/_gettsim/groupings.py:
```diff
@@ -93,8 +154,8 @@ def fg_id_numpy(  # noqa: PLR0913
                 child_hh_id == current_hh_id
-                # TODO (@hmgaudecker): Add correct conditions for grown up children
-                # https://github.com/iza-institute-of-labor-economics/gettsim/pulls/509
+                # TODO (@MImmesberger): Check correct conditions for grown up children
```
Good occasion to [check the remark over there](https://github.com/iza-institute-of-labor-economics/gettsim/pull/509#issuecomment-1948171381) and close that PR.

hmgaudecker commented on src/_gettsim/transfers/kindergeld.py:
```diff
@@ -1,11 +1,7 @@
 aggregate_by_group_kindergeld = {
-    "kumulativer_kindergeld_anspruch_tu": {
-        "source_col": "kindergeld_anspruch",
-        "aggr": "cumsum",
```
Please get rid of the cumsum aggregator then! (S+R for cumsum in entire project, remove all related code)

hmgaudecker reviewed (commented):
Amazing work, thanks a lot!!! 

Have to stop now at 29/1807 files... But there should be enough stuff to take care of already!

hmgaudecker commented on src/_gettsim/taxes/zu_verst_eink/freibetraege.py:
```diff
@@ -314,14 +318,14 @@ def eink_st_sonderausgaben_y_tu_mit_betreuung(
 def eink_st_abz_betreuungskost_y(
     eink_st_abzuege_params: dict,
-    betreuungskost_m: float,
+    betreuungskost_eltern_m: float,
```
```suggestion
    betreuungskost_eltern_y: float,
```

Although I think I'd prefer `betreuungskost_elternteil_y`, if I understand the variable correctly.

hmgaudecker commented on src/_gettsim/taxes/zu_verst_eink/freibetraege.py:
```diff
@@ -314,14 +318,14 @@ def eink_st_sonderausgaben_y_tu_mit_betreuung(
     ----------
-    betreuungskost_m
-        See basic input variable :ref:`betreuungskost_m <betreuungskost_m>`.
+    betreuungskost_eltern_m
```
```suggestion
    betreuungskost_eltern_y
```

hmgaudecker commented on src/_gettsim/taxes/zu_verst_eink/freibetraege.py:
```diff
@@ -314,14 +318,14 @@ def eink_st_sonderausgaben_y_tu_mit_betreuung(
-    betreuungskost_m
-        See basic input variable :ref:`betreuungskost_m <betreuungskost_m>`.
+    betreuungskost_eltern_m
+        See :func:`betreuungskost_eltern_m`.
```
```suggestion
        See :func:`betreuungskost_eltern_y`.
```

hmgaudecker commented on src/_gettsim/taxes/zu_verst_eink/freibetraege.py:
```diff
@@ -330,18 +334,18 @@ def eink_st_abz_betreuungskost_y(
     """
     out = min(
-        12 * betreuungskost_m,
+        12 * betreuungskost_eltern_m,
```
```suggestion
        betreuungskost_eltern_y,
```

hmgaudecker commented on src/_gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:
```diff
@@ -2,12 +2,12 @@
-    priv_rentenv_beitr_m_tu: float,
-    anz_erwachsene_tu: int,
+def vorsorgeaufw_alter_y_sn_einfuehrung(
+    ges_rentenv_beitr_m_sn: float,
```
Do the conversion outside of the function. Applies to lots of things in this module, it seems by glancing at the top. 

But feel free to open an issue for that, I thought we had converted all that at some point already.

MImmesberger replied:
I'll create an issue because this would entail rewriting some of the functions /
creating intermediate functions.

MImmesberger replied:
Most of the functions in this module only make sense on the aggregated level because
they involve Aufwendungen that are capped at the `sn` level. Of course,
we could split up these functions into a function that calculates Aufwendungen on the
individual level and a part that caps them at the `sn` level. But it seems like we
wouldn't gain much clarity in this module because the functions for which this would
work are rather short (e.g. the one you highlighted here).

hmgaudecker replied:
Sorry, I meant the conversion from monthly to annual!

MImmesberger replied:
Makes sense! Created #709

hmgaudecker commented on src/_gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:
```diff
@@ -32,72 +32,72 @@ def vorsorgeaufw_alter_y_tu_einfuehrung(
         eink_st_abzuege_params["einführungsfaktor_vorsorgeaufw_alter_ab_2005"]
-        * (2 * ges_rentenv_beitr_m_tu + priv_rentenv_beitr_m_tu)
-        - ges_rentenv_beitr_m_tu
+        * (2 * ges_rentenv_beitr_m_sn + priv_rentenv_beitr_m_sn)
```
Do we have the employer version separate, or do we hard-code that they are the same as employee contributions?

MImmesberger replied:
They are hard-coded to be the same as employee contributions which should be wrong for
Midijobs. I created #705.

MImmesberger commented on CHANGES.md:
```diff
@@ -8,6 +8,8 @@ releases are available on [Anaconda.org](https://anaconda.org/conda-forge/gettsi
 - {gh}`700` Data columns overwrite functions regardless of time unit
   ({ghuser}`lars-reimann`).
+- {gh}`694` Replace `tu` groupings with correct groupings. ({ghuser}`MImmesberger`,
+  \{ghuser} `lars-reimann`).
```
When you reviewed all the files, you undoubtedly deserve it.

MImmesberger commented on src/_gettsim/groupings.py:
```diff
@@ -5,13 +5,19 @@
 def create_groupings() -> dict[str, Callable]:
-    return {"bg_id": bg_id_numpy, "fg_id": fg_id_numpy, "sn_id": sn_id_numpy}
+    return {
+        "bg_id": bg_id_numpy,
```
I'll go with size in the future (`hh`, `fg`, `bg`, `eg`, `ehe`, `sn`).

hmgaudecker commented on CHANGES.md:
```diff
@@ -8,6 +8,8 @@ releases are available on [Anaconda.org](https://anaconda.org/conda-forge/gettsi
 - {gh}`700` Data columns overwrite functions regardless of time unit
   ({ghuser}`lars-reimann`).
+- {gh}`694` Replace `tu` groupings with correct groupings. ({ghuser}`MImmesberger`,
+  \{ghuser} `lars-reimann`, \{ghuser} `hmgaudecker`).
```
```suggestion
  {ghuser}`lars-reimann`, {ghuser}`hmgaudecker`).
```

hmgaudecker replied:
Are the pre-commit hooks messing this up? Or was it the extra whitespace?

MImmesberger replied:
Probably the whitespace together with the hooks.

hmgaudecker commented on src/_gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:
```diff
@@ -273,69 +273,69 @@ def vorsorgeaufw_y_tu_bis_2004(
 
     """
-    if gemeinsam_veranlagt_tu:
+    if anz_personen_sn > 1:
```
Remove if, do the formula with
```
0.5 = 1 / anz_personen_sn
2 = anz_personen_sn
```
This trick might apply at a bunch of places, actually. Feel free to open an issue if you do not want to tackle it here.

hmgaudecker commented on src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2.py:
```diff
@@ -243,8 +243,8 @@ def arbeitsl_geld_2_regelsatz_m_bg_bis_2010(
 
     """
     # Note that we, currently, do not support households with more than 2 adults.
-    weitere_erwachsene = max(anz_erwachsene_hh - 2, 0)
```
Is the comment above still relevant now that we have Bedarfsgemeinschaften? And how could we have additional grown-ups in a Bedarfsgemeinschaft?

MImmesberger replied:
So the law for this function says the following:

> Haben zwei Angehörige der Bedarfsgemeinschaft
> das 18. Lebensjahr vollendet, beträgt die Regelleistung
> jeweils 90 vom Hundert der Regelleistung nach Absatz 2.
> Die Regelleistung für sonstige erwerbsfähige Angehörige
> der Bedarfsgemeinschaft beträgt 80 vom Hundert der
> Regelleistung nach Absatz 2.

The additional grown-ups could be children of the parents. I don't see why the limitation to two adults in the Bedarfsgemeinschaft should be relevant anymore.

hmgaudecker replied:
Cool, thanks! I was just confused then.

hmgaudecker reviewed (commented):
Piecemeal once more, apologies. Great work, really!!!

hmgaudecker commented on src/_gettsim/transfers/kinderzuschl/kinderzuschl_eink.py:
```diff
@@ -21,15 +22,18 @@ def kinderzuschl_bruttoeink_eltern_m(
+    kindergeld_anspruch
+        See :func:`kindergeld_anspruch`.
+    erwachsen
+        See basic input variable :ref:`erwachsen <erwachsen>`.
```
This is a function. 

I think we should be able to get rid of the `kind` input variable now? Based on hh and elternteil pointers, we should be fine, no?

I find kind/erwachsen super confusing because it is so ambiguous. In the legal sense it is clear, age 18 is the border. But this somehow uses the SGB II definition (docstring says *Dependent child living with parents*), but we try to set that up in a more sensible way now?

hmgaudecker replied:
To some extent `kind` can actually be replaced by `kindergeld_anspruch`, I think. 

But we really need to check if different definitions apply in other areas of law (SGB II, whatever). But I think it often says "Kinder im Sinne des Bundeskindergeldgesetzes".

MImmesberger replied:
Yes definitely. But I think we should do this in a new PR.

For ALG2 Regelbedarf, "Erwachsen" seems to refer to anyone over 18 (independent of Kindergeld Anspruch). So there are definitely differences.

hmgaudecker commented on src/_gettsim/transfers/kinderzuschl/kinderzuschl_eink.py:
```diff
@@ -40,33 +44,36 @@ def kinderzuschl_bruttoeink_eltern_m(
 
     """
-    if eltern:
+    if erwachsen and (not kindergeld_anspruch):
```
Same as above

hmgaudecker commented on src/_gettsim/transfers/kinderzuschl/kinderzuschl_eink.py:
```diff
@@ -196,8 +203,9 @@ def kinderzuschl_eink_min_m_tu(
-    anz_kinder_hh
-        See basic input variable :ref:`anz_kinder_hh <anz_kinder_hh>`.
+    _kinderzuschl_anz_kinder_anspruch_bg
+        See basic input variable :ref:`_kinderzuschl_anz_kinder_anspruch_bg
```
```suggestion
        See :func:`_kinderzuschl_anz_kinder_anspruch_bg
```

hmgaudecker commented on src/_gettsim/transfers/kinderzuschl/kost_unterk.py:
```diff
@@ -1,96 +1,96 @@
+    anz_personen_bg: int, anz_personen_hh: int
 ) -> float:
-    """Calculate the share of tax units in household.
+    """Calculate the share of Bedarfsgemeinschaften in household.
```
```suggestion
    """Calculate the share of persons in this Bedarfsgemeinschaft among all persons in the household.
```

hmgaudecker commented on src/_gettsim/transfers/unterhaltsvors.py:
```diff
@@ -68,56 +107,73 @@ def unterhaltsvors_m(  # noqa: PLR0913
+    arbeitsl_geld_m: float,
 ) -> float:
-    """Calculate relevant income for advance on alimony payment on tax unit level.
+    """Calculate relevant income for advance on alimony payment on Steuernummer level.
```
```suggestion
    """Calculate relevant income for advance on alimony payment.
```
We're individual-level here, right?

hmgaudecker commented on src/_gettsim/transfers/wohngeld.py:
```diff
@@ -41,21 +49,21 @@ def wohngeld_m_hh(
     """Calculate housing benefit subtractions on the individual level.
 
-    Note that eink_st_y_tu is used as an approximation for taxes on income (as mentioned
+    Note that eink_st_y_sn is used as an approximation for taxes on income (as mentioned
```
This is actually a bad approximation as it may go systematically wrong for 1-earner couples (or secondary earner just doing a minijob) when first earner has sufficient income. I am not perfectly sure how this is being done in practice, though. Mayb open an issue with detailed Wohngeld questions?

MImmesberger replied:
Sorry, I can't quite follow here. Is this related to the definition of "Steuern vom Einkommen" (WoGG) or the grouping on `sn` level (because same `sn_id` does not imply same `hh_id`)?

hmgaudecker replied:
Yes, I thought the "Steuern vom Einkommen" was defined at the individual level. If there is only one earner with sufficient income, both will pay income tax. But I'd think that subtraction will happen only once.

MImmesberger replied:
Created #713

hmgaudecker commented on src/_gettsim/transfers/wohngeld.py:
```diff
@@ -496,9 +508,9 @@ def wohngeld_miete_m_hh_ab_2009(  # noqa: PLR0912 (see #516)
             i for i in wohngeld_params["heizkostenentlastung_m"] if isinstance(i, int)
         )
     if "heizkostenentlastung_m" in wohngeld_params:
```
Eventually get rid of these things via `@dates_active` -> issue.

hmgaudecker commented on src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2_eink.py:
```diff
@@ -4,9 +4,9 @@
-    eink_st_y_tu: float,
-    soli_st_y_tu: float,
-    anz_erwachsene_tu: int,
+    eink_st_y_sn: float,
```
```suggestion
    eink_st_m_sn: float,
```
etc.

hmgaudecker commented on src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2_eink.py:
```diff
@@ -4,9 +4,9 @@
+    anz_personen_sn: int,
     sozialv_beitr_m: float,
     arbeitsl_geld_2_eink_anr_frei_m: float,
     kind: bool,
```
```suggestion
```

The conditional expression can just go, I think. Kids have their own `sn` and we calculate income taxes for them -- all good.

hmgaudecker reviewed (commented):
I checked all code now and some very few tests. If there are any tests you think I should look at in particular, LMK. 

Really cool stuff!

MImmesberger commented on src/_gettsim/transfers/arbeitsl_geld_2/kost_unterk.py:
```diff
@@ -128,21 +128,21 @@ def _arbeitsl_geld_2_berechtigte_wohnfläche_bg(
     return min(wohnfläche_hh, maximum)
 
-    # if bewohnt_eigentum_hh and haushaltsgröße_hh < 5:
+    # if bewohnt_eigentum_hh and anz_personen_hh < 5:
```
No that comes from #425. I'll remove it.

MImmesberger commented on src/_gettsim/transfers/arbeitsl_geld_2/kost_unterk.py:
```diff
@@ -128,21 +128,21 @@ def _arbeitsl_geld_2_berechtigte_wohnfläche_bg(
     if bewohnt_eigentum_hh:
-        if haushaltsgröße_hh <= 4:
-            maximum = params[haushaltsgröße_hh]
+        if anz_personen_hh <= 4:
```
Yes, that's just an approximation until we know how to break down the Wohngeld variables to the individual level.

MImmesberger commented:
I don't think there are particular tests you have to look at.

Just so it doesn't sneak through the review process: In [1d78ded](https://github.com/iza-institute-of-labor-economics/gettsim/pull/694/commits/1d78ded3196731f422c3340349a1ffd3eaaac892), I set the minimum relevant income for ALG2 to 0. There were some test cases that resulted in negative values after getting rid of `tu`.

MImmesberger commented:
The rounding seems to be an issue when switching from `eink_st_y_sn / 12` to `eink_st_m_sn`, because rounding parameters are expected for the `eink_st_m_sn` key in the param yaml. They are specified for the `eink_st_y_sn` key only.

I'm not entirely sure what the expected behavior should be:
1. First round and then time convert (with optional rounding spec for converted column).
2. First time convert and then round (We could use a quick fix by specifying the `eink_st_m_sn` rounding spec in the params yaml).

Because `eink_st_m_sn` is used as an approximation for Lohnsteuer (at least for Elterngeld and Grundsicherung; not sure about ALG2) we cannot infer expected behavior from the law.

hmgaudecker commented:
> Just so it doesn't sneak through the review process: In [1d78ded](https://github.com/iza-institute-of-labor-economics/gettsim/pull/694/commits/1d78ded3196731f422c3340349a1ffd3eaaac892), I set the minimum relevant income for ALG2 to 0. There were some test cases that resulted in negative values after getting rid of `tu`.

Cool, thanks for the pointer. In principle, I think that ALG II should guarantee a consumption floor, but it might be that the mechanism for that is more complicated. I would have no idea. But I'd kind of like to guarantee that subsistence level (thinking of running GETTSIM on models that do require a positive consumption), unless the law is different.

If possible, I'd prefer to revert and make an issue that we'll need to investigate this case.

hmgaudecker commented:
> The rounding seems to be an issue when switching from `eink_st_y_sn / 12` to `eink_st_m_sn`, because rounding parameters are expected for the `eink_st_m_sn` key in the param yaml. They are specified for the `eink_st_y_sn` key only.
> 
> I'm not entirely sure what the expected behavior should be:
> 
>     1. First round and then time convert (with optional rounding spec for converted column).
> 
>     2. First time convert and then round (We could use a quick fix by specifying the `eink_st_m_sn` rounding spec in the params yaml).
> 
> 
> Because `eink_st_m_sn` is used as an approximation for Lohnsteuer (at least for Elterngeld and Grundsicherung; not sure about ALG2) we cannot infer expected behavior from the law.

Behavior 1. and the last two cases we should get rid of eventually now that we have Lohnsteuer.

MImmesberger commented:
> > The rounding seems to be an issue when switching from `eink_st_y_sn / 12` to `eink_st_m_sn`, because rounding parameters are expected for the `eink_st_m_sn` key in the param yaml. They are specified for the `eink_st_y_sn` key only.
> > I'm not entirely sure what the expected behavior should be:
> > ```
> > 1. First round and then time convert (with optional rounding spec for converted column).
> > 
> > 2. First time convert and then round (We could use a quick fix by specifying the `eink_st_m_sn` rounding spec in the params yaml).
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Because `eink_st_m_sn` is used as an approximation for Lohnsteuer (at least for Elterngeld and Grundsicherung; not sure about ALG2) we cannot infer expected behavior from the law.
> 
> Behavior 1. and the last two cases we should get rid of eventually now that we have Lohnsteuer.

@lars-reimann Can you have a look?

hmgaudecker commented:
> @lars-reimann Can you have a look?

Would be great, but let's do that separately from here. Apologies if that means reverting something I noted. Also fine to skip a couple of tests for the moment.

MImmesberger commented:
> Cool, thanks for the pointer. In principle, I think that ALG II should guarantee a consumption floor, but it might be that the mechanism for that is more complicated. I would have no idea. But I'd kind of like to guarantee that subsistence level (thinking of running GETTSIM on models that do require a positive consumption), unless the law is different.

I think we are talking about different things here. I realized that the issue I raised is solved once we fix #606 because it makes sure that we don't compute negative "anrechenbares Erwerbseinkommen" on the individual level.

hmgaudecker commented on src/_gettsim/transfers/kinderzuschl/kinderzuschl_eink.py:
```diff
@@ -21,15 +22,18 @@ def kinderzuschl_bruttoeink_eltern_m(
     """
 
-    if eltern:
+    if erwachsen and (not kindergeld_anspruch):
```
Can you make a note of this somewhere? My point is that we simply should think about some value other than zero for people who do not have potentially eligible children. Probably tackle in #704, but we should remember explicitly to change this condition.

MImmesberger replied:
I think once we implemented #708, we can get rid of the condition entirely but let's see. I added a ToDo note and referenced it in both issues.

hmgaudecker reviewed (approved):
Great work!!! Thanks a ton.
